subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -58,96 +59,109 @@
  *   - remember the current watched root project directories as old root directories for the next build
  *   - remove all non-watched root project directories from the old root directories.
  */
-public class HierarchicalFileWatcherUpdater implements FileWatcherUpdater {
+public class HierarchicalFileWatcherUpdater extends AbstractFileWatcherUpdater {
     private static final Logger LOGGER = LoggerFactory.getLogger(HierarchicalFileWatcherUpdater.class);
 
-    private final Multimap<String, Path> trackedDirectoriesForSnapshot = HashMultimap.create();
-
     private final Set<Path> watchedHierarchies = new HashSet<>();
+    private final FileWatcher fileWatcher;
 
-    private final Set<Path> knownRootProjectDirectoriesFromCurrentBuild = new HashSet<>();
-    private final Set<Path> watchedRootProjectDirectoriesFromPreviousBuild = new HashSet<>();
-
-    private final FileWatcher watcher;
     private final FileSystemLocationToWatchValidator locationToWatchValidator;
 
-    public HierarchicalFileWatcherUpdater(FileWatcher watcher, FileSystemLocationToWatchValidator locationToWatchValidator) {
-        this.watcher = watcher;
+    private final Set<Path> hierarchiesToWatchFromCurrentBuild = new HashSet<>();
+    private final Set<Path> watchedHierarchiesFromPreviousBuild = new HashSet<>();
```

Comment:
"watcheable" vs "watched" sounds like a good contrast to me here as well (wrt renaming `ToWatch` to `Watchable`).

Reply:
The ones from the previous build actually are `watched` hierarchies, since we only pass on the ones we watched at the end of the build to the next build.

Reply:
Exactly.

Reply:
Renamed.